### PR TITLE
*8348* Convert category grid to listbuilder

### PR DESCRIPTION
**Pull request by asmecher**

### CODE REVIEW

File: classes/journal/categories/CategoryDAO.inc.php [hunk 1]
```diff
@@ -36,13 +36,19 @@ function getCategories() {
 	}
 
 	/**
-	 * Rebuild the cache.
+	 * Get an iterator of category objects.
 	 */
```
Comment (beghelli) [inline classes/journal/categories/CategoryDAO.inc.php]: I think you can use @return DAOResultFactory right?

File: classes/journal/categories/CategoryDAO.inc.php [hunk 2]
```diff
@@ -36,13 +36,19 @@ function getCategories() {
...
-	function rebuildCache() {
-		// Read the full set of categories into an associative array
+	function getIterator() {
```
Comment (beghelli) [inline classes/journal/categories/CategoryDAO.inc.php]: Also, it's used anywhere else outside the class? If not, we can hide it declaring it protected

File: controllers/listbuilder/admin/categories/CategoryListbuilderGridCellProvider.inc.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	//
+	// Template methods from GridCellProvider
+	//
+	/**
+	 * @copydoc GridCellProvider::getTemplateVarsFromRowColumn()
+	 */
+	function getTemplateVarsFromRowColumn($row, $column) {
+		$category =& $row->getData(); /* @var $category Category */
```
Comment (beghelli) [inline controllers/listbuilder/admin/categories/CategoryListbuilderGridCellProvider.inc.php]: is this reference necessary?

File: controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+class CategoryListbuilderHandler extends ListbuilderHandler {
+
+	/**
+	 * Constructor
+	 */
+	function CategoryListbuilderHandler() {
+		parent::ListbuilderHandler();
+		$this->addRoleAssignment(
+			ROLE_ID_SITE_ADMIN,
```
Comment (beghelli) [inline controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php]: not journal manager?

File: controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			array('fetch', 'fetchRow', 'save', 'fetchOptions')
+		);
+	}
+
+
+	//
+	// Overridden template methods
+	//
+	/**
+	 * @copydoc SetupListbuilderHandler::initialize()
```
Comment (beghelli) [inline controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php]: wrong doc

File: controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		$titleColumn->setCellProvider(new CategoryListbuilderGridCellProvider());
+		$this->addColumn($titleColumn);
+	}
+
+	/**
+	 * @copydoc GridHandler::loadData()
+	 */
+	function loadData($request) {
+		$categoryDao = DAORegistry::getDAO('CategoryDAO');
+		return $categoryDao->getIterator();
```
Comment (beghelli) [inline controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php]: gotcha, public then.

File: controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	/**
+	 * @copydoc GridHandler::getRowDataElement
+	 */
+	function getRowDataElement($request, &$rowId) {
```
Comment (beghelli) [inline controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php]: $rowId needs the reference?

File: controllers/tab/admin/categories/form/CategorySettingsForm.inc.php [hunk 8]
```diff
@@ -53,6 +53,55 @@ function readInputData() {
 	function execute($request) {
 		$site = $request->getSite();
 		$site->updateSetting('categoriesEnabled', (int) $this->getData('categoriesEnabled'));
+		ListbuilderHandler::unpack($request, $this->getData('categories'));
+
+	}
+
+	/**
+	 * @copydoc ListbuilderHandler::insertEntry()
```
Comment (beghelli) [inline controllers/tab/admin/categories/form/CategorySettingsForm.inc.php]: can we copydoc from a method we are actually not extending?

File: controllers/tab/settings/masthead/form/MastheadForm.inc.php [hunk 9]
```diff
@@ -122,7 +122,7 @@ function updateEntry($request, $rowId, $newRowId) {
 	 * @copydoc ListbuilderHandler::deleteEntry()
 	 */
 	function deleteEntry($request, $rowId) {
-		if (isset($this->categories[$rowId['name']])) unset($this->categories[$rowId['name']]);
```
Comment (beghelli) [inline controllers/tab/settings/masthead/form/MastheadForm.inc.php]: this was working before?

File: controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php (hunk 4 above)
Comment (asmecher) [inline controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php]: No -- this grid is to maintain the list of available categories; there's another listbuilder to assign them to journals, which is the JM's purview.

File: controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php (hunk 7 above)
Comment (asmecher) [inline controllers/listbuilder/admin/categories/CategoryListbuilderHandler.inc.php]: Yes; see https://github.com/pkp/pkp-lib/commit/e73998ccd29d4525634937d47c2430454ef2762c

File: controllers/tab/admin/categories/form/CategorySettingsForm.inc.php (hunk 8 above)
Comment (asmecher) [inline controllers/tab/admin/categories/form/CategorySettingsForm.inc.php]: Yes, Doxygen doesn't care.

File: controllers/tab/settings/masthead/form/MastheadForm.inc.php (hunk 9 above)
Comment (asmecher) [inline controllers/tab/settings/masthead/form/MastheadForm.inc.php]: No, I fixed it as part of this.

**Pull request merged**